src/jdk.compiler/share/classes/com/sun/tools/javac/comp/ThisEscapeAnalyzer.java

```diff
@@ -0,0 +1,1475 @@
+            return;
+
+        // Sanity check
+        Assert.check(this.checkInvariants(true, false));
+
+        // Can this expression node possibly leave a 'this' reference on the stack?
+        final boolean referenceExpressionNode;
+        switch (tree.getTag()) {
+        case CASE:
+        case SWITCH_EXPRESSION:
+        case CONDEXPR:
+        case YIELD:
+        case APPLY:
+        case NEWCLASS:
+        case NEWARRAY:
+        case LAMBDA:
+        case PARENS:
+        case ASSIGN:
+        case TYPECAST:
+        case INDEXED:
+        case SELECT:
+        case REFERENCE:
+        case IDENT:
+        case NULLCHK:
+        case LETEXPR:
+            referenceExpressionNode = true;
+            break;
+        default:
+            referenceExpressionNode = false;
+            break;
+        }
+
+        // Scan node
+        super.scan(tree);
+
+        // Sanity check
+        Assert.check(this.checkInvariants(true, referenceExpressionNode));
+
+        // Discard any direct 'this' reference that's incompatible with the target type
+        if (referenceExpressionNode) {
+
+            // We treat instance methods as having a "value" equal to their instance
+            Type type = tree.type;
+            final Symbol sym = TreeInfo.symbolFor(tree);
+            if (sym != null &&
+                sym.kind == MTH &&
+                (sym.flags() & Flags.STATIC) == 0) {
+                type = sym.owner.type;
+            }
+
+            // If the expression type is incompatible with 'this', discard it
+            if (type != null && !this.isSubtype(this.targetClass.sym.type, type))
+                this.refs.remove(ExprRef.direct(this.depth));
+        }
+    }
+
+//
+// Visitor methods - Class Declarations
+//
+
+    @Override
+    public void visitClassDef(JCClassDecl tree) {
+        return;     // we're busy analyzing another class - skip
+    }
+
+//
+// Visitor methods - Variable Declarations
+//
+
+    @Override
+    public void visitVarDef(JCVariableDecl tree) {
+
+        // Skip if ignoring warnings for this variable
+        if (this.suppressed.contains(tree.sym))
+            return;
+
+        // Scan initializer, if any
+        this.scan(tree.init);
+        if (this.isParamOrVar(tree.sym))
+            this.refs.replaceExprs(this.depth, direct -> new VarRef(tree.sym, direct));
+        else
+            this.refs.discardExprs(this.depth);         // we don't track fields yet
+    }
+
+//
+// Visitor methods - Methods
+//
+
+    @Override
+    public void visitMethodDef(JCMethodDecl tree) {
+        Assert.check(false);        // we should never get here
+    }
+
+    @Override
+    public void visitApply(JCMethodInvocation invoke) {
+
+        // Get method symbol
+        final MethodSymbol sym = (MethodSymbol)TreeInfo.symbolFor(invoke.meth);
+
+        // Recurse on method expression
+        this.scan(invoke.meth);
+        final boolean direct = this.refs.remove(ExprRef.direct(this.depth));
```

Comment:
> This will then determine how to interpret this in the context of that method analysis - e.g. when we see a JCIdent for this, we create a direct/indirect ExprRef based on what the receiver kind was. Correct?

Yes, exactly.

When we "invoke" a method, we "preload" the set of current references that the method is going to have when it starts. That "preload" step includes any references from (a) the method receiver (non-static methods only) and (b) method parameters.

To the extent the method receiver has a direct/indirect reference, that turns into a direct/indirect `ThisRef` during method execution. You can see this happen in the very next lines after what you quoted. But of course the method invocation (the "apply") could have been applied to any arbitrary expression as the receiver.